drivers/i2s/i2s_silabs_siwx91x.c

```diff
@@ -0,0 +1,908 @@
+		key = irq_lock();
+		if (stream->state != I2S_STATE_RUNNING) {
+			irq_unlock(key);
+			return -EIO;
+		}
+
+		stream->state = I2S_STATE_STOPPING;
+		irq_unlock(key);
+		stream->last_block = true;
+		break;
+
+	case I2S_TRIGGER_DRAIN:
+		key = irq_lock();
+		if (stream->state != I2S_STATE_RUNNING) {
+			irq_unlock(key);
+			return -EIO;
+		}
+
+		stream->state = I2S_STATE_STOPPING;
+		irq_unlock(key);
+		break;
+
+	case I2S_TRIGGER_DROP:
+		if (stream->state == I2S_STATE_NOT_READY) {
+			return -EIO;
+		}
+
+		i2s_siwx91x_stream_disable(stream, stream->dma_dev);
+		stream->queue_drop(stream);
+		stream->state = I2S_STATE_READY;
+		break;
+
+	case I2S_TRIGGER_PREPARE:
+		if (stream->state != I2S_STATE_ERROR) {
+			return -EIO;
+		}
+
+		stream->state = I2S_STATE_READY;
+		stream->queue_drop(stream);
+		break;
+
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int i2s_siwx91x_init(const struct device *dev)
+{
+	const struct i2s_siwx91x_config *cfg = dev->config;
+	struct i2s_siwx91x_data *data = dev->data;
+	int ret;
+
+	ret = clock_control_on(cfg->clock_dev, cfg->clock_subsys_peripheral);
+	if (ret) {
+		return ret;
+	}
+
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret) {
+		return ret;
+	}
+
+	cfg->reg->I2S_IER_b.IEN = 1;
+	cfg->reg->I2S_IRER_b.RXEN = 0;
+	cfg->reg->I2S_ITER_b.TXEN = 0;
+
+	k_sem_init(&data->rx.sem, 0, CONFIG_I2S_SILABS_SIWX91X_RX_BLOCK_COUNT);
+	k_sem_init(&data->tx.sem, CONFIG_I2S_SILABS_SIWX91X_TX_BLOCK_COUNT,
+		   CONFIG_I2S_SILABS_SIWX91X_TX_BLOCK_COUNT);
+
+	return ret;
+}
+
+static DEVICE_API(i2s, i2s_siwx91x_driver_api) = {
+	.configure = i2s_siwx91x_configure,
+	.config_get = i2s_siwx91x_config_get,
+	.read = i2s_siwx91x_read,
+	.write = i2s_siwx91x_write,
+	.trigger = i2s_siwx91x_trigger,
+};
+
+#define SIWX91X_I2S_INIT(inst)                                                                     \
+	PINCTRL_DT_INST_DEFINE(inst);                                                              \
+	struct i2s_siwx91x_queue_item                                                              \
+		rx_ring_buf_##inst[CONFIG_I2S_SILABS_SIWX91X_RX_BLOCK_COUNT + 1];                  \
+	struct i2s_siwx91x_queue_item                                                              \
+		tx_ring_buf_##inst[CONFIG_I2S_SILABS_SIWX91X_TX_BLOCK_COUNT + 1];                  \
+                                                                                                   \
+	BUILD_ASSERT(((I2S0_Type *)DT_INST_REG_ADDR(inst) != I2S0) ||                              \
+			     (DT_INST_PROP(inst, silabs_channel_group) < 2),                       \
+		     "Invalid channel group!");                                                    \
+	BUILD_ASSERT(((I2S0_Type *)DT_INST_REG_ADDR(inst) != I2S1) ||                              \
+			     (DT_INST_PROP(inst, silabs_channel_group) < 1),                       \
+		     "Invalid channel group!");                                                    \
```

Comment:
The DT already describe the various instances of the hardware and the limitations of each instances. I believe this is not the job of the driver to validate the DT. Typically, we currently not check the consistency of `reg`, `clock` or `pinctrl` properties. 

So, I think we should get rid of these `BUILD_ASSERT()` since they refer to specific instances of the driver.

@jhedberg, @asmellby, do you have another mind?